library/ecp.c

```diff
@@ -767,6 +767,68 @@ int mbedtls_ecp_point_write_binary( const mbedtls_ecp_group *grp,
     return( ret );
 }
 
+#if defined(MBEDTLS_ECP_SHORT_WEIERSTRASS_ENABLED)
+static int mbedtls_ecp_sw_derive_y( const mbedtls_ecp_group *grp,
+                                    const mbedtls_mpi *X,
+                                    mbedtls_mpi *Y,
+                                    int parity_bit )
+{
+    /* y^2 = x^3 + ax + b
+     * sqrt(w) = w^((p+1)/4) mod p   (for prime p where p = 3 mod 4)
+     *
+     * Note: this method for extracting square root does not validate that w
+     * was indeed a square so this function will return garbage in Y if X
+     * does not correspond to a point on the curve.
+     */
+
+    /* Check prerequisite p = 3 mod 4 */
+    if( mbedtls_mpi_get_bit( &grp->P, 0 ) != 1 ||
+        mbedtls_mpi_get_bit( &grp->P, 1 ) != 1 )
+        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+
+    int ret;
+    mbedtls_mpi exp;
+    mbedtls_mpi_init(&exp);
+
+    /* use Y to store intermediate results */
+    /* y^2 = x^3 + ax + b = (x^2 + a)x + b */
+    /* x^2 */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( Y, X, X ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( Y, Y, &grp->P ) );
```

Comment:
@mprse: This was briefly discussed above in https://github.com/Mbed-TLS/mbedtls/pull/6282#issuecomment-1309252158
The code is correct with the modulo `P` operation after each step.  I also asked in the link whether it might be safe (in all valid use cases) to remove some of those operations.  For use with known curves, I think it is safe to remove the module between the mul and add of `&grp->A` and `&grp->B`.  I do not think modulo `P` is safe in all cases to omit between multiplies or exponentiation.  Thoughts, @mpg?

Reply:
From a purely correctness perspective, it would indeed be OK to only reduce at the end. However, omitting the reduction after multiplication would be bad for performance, and even RAM usage, as it means the intermediate values would grow in size: if `grp->P` is x limbs, then after y multiplications the result is x*y limbs. While if we reduce after each multiplication, we keep the size under control at x limbs.

For addition, things are not as bad, if `grp->P` is p bits, then after q additions the result is at most p+q bits. But still, reducing after each operation is a bit cleaner, it keeps the result in the range `[0, grp->P)`. (Note that we are in the process of developing an API for modular operations, that will in the future be used everywhere in `ecp.c`, which will always reduce at the same time as doing the operation, and will eventually do all operations in constant time.)

Actually, there's already some kind of "modular operations" interface in `ecp.c`, currently meant to be used through the macros `MPI_ECP_ADD()`, ` MPI_ECP_MUL()`, [etc](https://github.com/Mbed-TLS/mbedtls/blob/development/library/ecp.c#L1129). In addition to reducing code clutter, it would improve performance, by making use of fast reduction functions for curves that have them (while `mbedtls_mpi_mod_mpi()` is not curve-aware and is a generic slow version), and would probably facilitate future changes by being more uniform with the rest of the code.